From 8ca5507b64b0d98b538dd19d51f3189a994f3721 Mon Sep 17 00:00:00 2001
From: Purushottam Narayana <purushottam.narayana@lge.com>
Date: Wed, 10 Jan 2018 14:55:24 -0800
Subject: [PATCH] Fix scaling not working for specific display windows.

:Release Notes:
Fix scaling not working for specific display windows

:Detailed Notes:
Some of the coordinates considered invalid seems to be valid ones in YT.

:Testing Performed:
Play youtube in webbrowser and confirm it fits correctly.

:QA Notes:

:Issues Addressed:
[PLAT-48284] Video is not displayed well on youtube site in web browser app
---
 drivers/gpu/drm/drm_plane.c | 11 +++++++----
 1 file changed, 7 insertions(+), 4 deletions(-)

diff --git a/drivers/gpu/drm/drm_plane.c b/drivers/gpu/drm/drm_plane.c
index fd04078..0f6ba41 100644
--- a/drivers/gpu/drm/drm_plane.c
+++ b/drivers/gpu/drm/drm_plane.c
@@ -497,7 +497,7 @@ static int __setplane_internal(struct drm_plane *plane,
 	}
 
 	/* BRCM RPI drivers gives div 0 exception when display width/height being 0 */
-	if (!(crtc_w & crtc_h)) {
+	if (crtc_w == 0 || crtc_h == 0) {
 		DRM_DEBUG_KMS("Invalid CRTC coordinates %ux%u+%d+%d\n",
 			      crtc_w, crtc_h, crtc_x, crtc_y);
 		ret = -ERANGE;
@@ -676,7 +676,9 @@ int drm_mode_plane_set_custom_prop(struct drm_device *dev, struct drm_plane *pla
 				   scale_param->crtc_w, scale_param->crtc_h);
 
 			/* BRCM RPI drivers gives div 0 exception when display width/height being 0 */
-			if (!scale_param || !(scale_param->crtc_w & scale_param->crtc_h)) {
+			/* Scale param check is not required anymore as stack address passing is rectified in
+			 upper layer */
+			if (scale_param->crtc_w == 0 || scale_param->crtc_h == 0) {
 				DRM_DEBUG_KMS("\n Invalid values, Dst rect width or height or both zeros");
 				return -ENOENT;
 			}
@@ -692,6 +694,7 @@ int drm_mode_plane_set_custom_prop(struct drm_device *dev, struct drm_plane *pla
 				plane->state->src_y = scale_param->src_y;
 				plane->state->src_w = scale_param->src_w << 16;
 				plane->state->src_h = scale_param->src_h << 16;
+				ret = 0;
 			} else {
 				ret = __setplane_internal_custom(plane, crtc, fb, scale_param->crtc_x, scale_param->crtc_y,
 												 scale_param->crtc_w, scale_param->crtc_h,
@@ -724,12 +727,12 @@ int drm_mode_plane_set_custom_prop(struct drm_device *dev, struct drm_plane *pla
 				else
 					return -ENOENT;
 			}
-			if (!(plane->state->crtc_w & plane->state->crtc_h)) {
+			if (plane->state->crtc_w == 0 || plane->state->crtc_h == 0) {
 				plane->state->crtc_w = fb->width;
 				plane->state->crtc_h = fb->height;
 				DRM_DEBUG_KMS("\n Dst width or height of dst is 0");
 			}
-			if (!(plane->state->src_w & plane->state->src_h)) {
+			if (plane->state->src_w == 0 || plane->state->src_h == 0) {
 				/* src_{x,y,w,h} values are 16.16 fixed point */
 				plane->state->src_w = fb->width << 16;
 				plane->state->src_h = fb->height << 16;
-- 
1.9.1

